index.js

```diff
@@ -1,167 +1,211 @@
+    return this
+  }
+
+  subtract (other) {
+    this.x -= other.x
+    this.y -= other.y
+    this.z -= other.z
+    return this
+  }
+
+  plus (other) {
+    return this.offset(other.x, other.y, other.z)
+  }
+
+  minus (other) {
+    return this.offset(-other.x, -other.y, -other.z)
+  }
+
+  scaled (scalar) {
+    return new Vec3(this.x * scalar, this.y * scalar, this.z * scalar)
+  }
+
+  abs () {
+    return new Vec3(Math.abs(this.x), Math.abs(this.y), Math.abs(this.z))
+  }
+
+  volume () {
+    return this.x * this.y * this.z
+  }
+
+  modulus (other) {
+    return new Vec3(
+      euclideanMod(this.x, other.x),
+      euclideanMod(this.y, other.y),
+      euclideanMod(this.z, other.z))
+  }
+
+  distanceTo (other) {
+    var dx = other.x - this.x
+    var dy = other.y - this.y
+    var dz = other.z - this.z
+    return Math.sqrt(dx * dx + dy * dy + dz * dz)
+  }
+
+  equals (other) {
+    return this.x === other.x && this.y === other.y && this.z === other.z
+  }
+
+  toString () {
+    return '(' + this.x + ', ' + this.y + ', ' + this.z + ')'
+  }
 
-function Vec3(x, y, z) {
-  this.x = x;
-  this.y = y;
-  this.z = z;
+  clone () {
+    return this.offset(0, 0, 0)
+  }
+
+  min (other) {
+    return new Vec3(Math.min(this.x, other.x), Math.min(this.y, other.y), Math.min(this.z, other.z))
+  }
+
+  max (other) {
+    return new Vec3(Math.max(this.x, other.x), Math.max(this.y, other.y), Math.max(this.z, other.z))
+  }
+
+  norm () {
+    return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z)
+  }
+
+  dot (other) {
+    return this.x * other.x + this.y * other.y + this.z * other.z
+  }
+
+  cross (other) {
+    return new Vec3(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z, this.x * other.y - this.y * other.x)
+  }
+
+  unit () {
+    const norm = this.norm()
+    if (norm === 0) {
+      return this.clone()
+    } else {
+      return this.scaled(1 / norm)
+    }
+  }
+
+  normalize () {
+    const norm = this.norm()
+    if (norm !== 0) {
+      this.x /= norm
+      this.y /= norm
+      this.z /= norm
+    }
+    return this
+  }
+
+  scale (scalar) {
+    this.x *= scalar
+    this.y *= scalar
+    this.z *= scalar
+    return this
+  }
+
+  xyDistanceTo (other) {
+    var dx = other.x - this.x
+    var dy = other.y - this.y
+    return Math.sqrt(dx * dx + dy * dy)
+  }
+
+  xzDistanceTo (other) {
+    var dx = other.x - this.x
+    var dz = other.z - this.z
+    return Math.sqrt(dx * dx + dz * dz)
+  }
+
+  yzDistanceTo (other) {
+    var dy = other.y - this.y
+    var dz = other.z - this.z
+    return Math.sqrt(dy * dy + dz * dz)
+  }
+
+  innerProduct (other) {
+    return this.x * other.x + this.y * other.y + this.z * other.z
+  }
+
+  manhattanDistanceTo (other) {
+    return Math.abs(other.x - this.x) + Math.abs(other.y - this.y) + Math.abs(other.z - this.z)
+  }
+
+  toArray () {
+    return [this.x, this.y, this.z]
+  }
 }
 
-function v(x, y, z) {
+function v (x, y, z) {
```

Comment:
I'm thinking about releasing a 1.0.0 of Vec3 instead of these 0.x version, and I think we could remove this and expose directly Vec3 in the process, what do you think ?

Reply:
I agree! I left it for now, because I didn't wanted to break everything